launchlib/launcher.go

```diff
@@ -33,44 +33,59 @@ const (
-	Primary      *exec.Cmd
-	SubProcesses map[string]*exec.Cmd
+type CmdWithOutputFileName struct {
+	Cmd            *exec.Cmd
+	OutputFileName string
 }
 
-func CompileCmdsFromConfigFiles(staticConfigFile, customConfigFile string, stdout io.Writer) (ServiceCommands, error) {
-	staticConfig, customConfig, err := GetConfigsFromFiles(staticConfigFile, customConfigFile, stdout)
-	if err != nil {
-		return ServiceCommands{}, err
-	}
-	return CompileCmdsFromConfig(&staticConfig, &customConfig, stdout)
+type ServiceCmds struct {
+	Primary  CmdWithOutputFileName
+	SubProcs map[string]CmdWithOutputFileName
 }
 
-func CompileCmdsFromConfig(staticConfig *PrimaryStaticLauncherConfig, customConfig *PrimaryCustomLauncherConfig, stdout io.Writer) (ServiceCommands, error) {
-	cmds := ServiceCommands{
-		SubProcesses: make(map[string]*exec.Cmd),
+func CompileCmdsFromConfig(staticConfig *PrimaryStaticLauncherConfig, customConfig *PrimaryCustomLauncherConfig,
+	stdout io.Writer) (serviceCmds *ServiceCmds, rErr error) {
+	serviceCmds = &ServiceCmds{
+		SubProcs: make(map[string]CmdWithOutputFileName),
 	}
+
 	var err error
-	if cmds.Primary, err = CompileCmdFromConfig(&staticConfig.StaticLauncherConfig, &customConfig.CustomLauncherConfig, stdout); err != nil {
-		return ServiceCommands{}, err
+	primaryCmd, err := compileCmdFromConfig(&staticConfig.StaticLauncherConfig, &customConfig.CustomLauncherConfig,
+		stdout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compile command for primary command")
 	}
-
+	serviceCmds.Primary = CmdWithOutputFileName{Cmd: primaryCmd, OutputFileName: PrimaryOutputFile}
 	for name, subProcStatic := range staticConfig.SubProcesses {
 		subProcCustom, ok := customConfig.SubProcesses[name]
 		if !ok {
-			return ServiceCommands{}, errors.Errorf("no custom launcher config exists for subProcess config '%s'", name)
+			return nil, errors.Errorf("no custom launcher config exists for subProcess config '%s'", name)
+		}
+		subProcOutputFileName := fmt.Sprintf(SubProcessOutputFileFormat, name)
+		subProcStdout, err := os.Create(subProcOutputFileName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create subProcess output file: %s",
+				subProcOutputFileName)
 		}
-		if cmds.SubProcesses[name], err = CompileCmdFromConfig(&subProcStatic, &subProcCustom, stdout); err != nil {
-			return ServiceCommands{}, errors.Wrapf(err, "unable to compile command for subProcess config '%s'", name)
+		defer func() {
+			if cErr := subProcStdout.Close(); rErr == nil && cErr != nil {
+				rErr = errors.Wrapf(err, "failed to close subProcess output file: %s",
+					subProcOutputFileName)
+			}
+		}()
+		subProc, err := compileCmdFromConfig(&subProcStatic, &subProcCustom, subProcStdout)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to compile command for subProcess config '%s'", name)
 		}
+		serviceCmds.SubProcs[name] = CmdWithOutputFileName{Cmd: subProc, OutputFileName: subProcOutputFileName}
```

Comment:
Could we split the subprocess compilation part into it's own function?  Since we are also opening and closing the log file it makes this for loop pretty lengthy.